scriptmodules/supplementary/emulationstation/configscripts/sdl2.sh

```diff
@@ -0,0 +1,178 @@
+       axis)
+           if [[ $input_value == "-1" ]]; then
+               echo "$sdl2_mapped_input:-a${input_id}" >> /tmp/sdl2.txt
+           else
+               echo "$sdl2_mapped_input:+a${input_id}" >> /tmp/sdl2.txt
+           fi
+           ;;
+       hat)
+           echo "$sdl2_mapped_input:hat${input_id}.${input_value}" >> /tmp/sdl2.txt
+           ;;
+       button)
+           echo "$sdl2_mapped_input:b${input_id}" >> /tmp/sdl2.txt
+           ;;
+       *)
+           ;;
+     esac
+}
+
+function onend_sdl2_joystick() {
+    # check whether SDL2 already has a mapping for this GUID
+    if [[ "$(_check_gamepad_sdl2)" == "MAPPED" ]]; then
+        echo "W: Device \"$DEVICE_NAME\" (guid: $DEVICE_GUID) is already known by SDL2, gamecontroller mapping was not created"
+        return
+    fi
+
+    # gamecontroller name sanitization:
+    # - replace unexpected chars with '-'
+    # - trim blanks from the name (beginning/end)
+    # - replace ',' with space, since ',' is a delimiter in the mapping string
+    local joyname="$(echo ${DEVICE_NAME//[:><?\"\/\\|*]/-} | tr -s '[:blank:]' | tr ',' ' ')"
+    local select_value
+    local hotkey_value
+    local mapping
+
+    # add each mapping value and save the values for Select/Back Hotkey/Guide
+    # don't add the Hotkey/Guide mapping yet
+    while read m; do
+       if [[ $m == "guide:*" ]]; then
+            hotkey_value=$m;
+            continue;
+        fi
+
+        if [[ $m == "back:*" ]]; then
+            select_value=$m;
+        fi
+
+        mapping+="$m,"
+     done < <(sort /tmp/sdl2.txt)
+
+    # add the mapping for Hotkey as Guide IIF it's different than Select
+    if [[ ($select_value != $hotkey_value) && (-n "$hotkey_value") ]]; then
+        mapping+="$hotkey_value,"
+    fi
+    local sdl_configs="$configdir/all/sdl2_gamecontrollerdb.txt"
+    mapping="${DEVICE_GUID},${joyname},platform:Linux,${mapping}"
+
+    [[ -n $__debug ]] && \
+        echo "Mapping is $mapping"
+
+    # update the mapping when it's present, otherwise append it to the file
+    if grep --silent --no-messages ${DEVICE_GUID} "$sdl_configs"; then
+        sed -i "s/^${DEVICE_GUID}.*$/$mapping/" "$sdl_configs"
+    else
+        echo $mapping >> "$sdl_configs"
+    fi
+    rm -f /tmp/sdl2.txt
+}
+
+# function to check whether a gamepad GUID is already in SDL2's mapping list
+# we need to load SDL2's GameController subsystem and query by GUID
+function _check_gamepad_sdl2() {
+
+cat << EOF > "/tmp/guid_check.py"
+from sdl2 import SDL_GameControllerMappingForGUID,\
+    SDL_Init,SDL_Quit,SDL_Error,\
+    SDL_INIT_GAMECONTROLLER
+from sdl2.joystick import SDL_JoystickGetGUIDFromString
+
+if SDL_Init(SDL_INIT_GAMECONTROLLER) < 0:
+    exit(2)
+
+guid_str=b"${DEVICE_GUID}"
+guid = SDL_JoystickGetGUIDFromString(guid_str)
+map = SDL_GameControllerMappingForGUID(guid)
+SDL_Quit()
+
+if map is not None:
+    exit(0)
+
+exit(1)
+EOF
+    python3 /tmp/guid_check.py 1>/dev/null 2>&1 && echo "MAPPED"
```

Comment:
/dev/shm would be quicker perhaps ? Or just include the python in its own file and pass it arguments ?

Reply:
Could be, but it's not something that's executed regularly, I think `/tmp` is fine. I wanted to avoid adding an extra file just for this check.